src/meltano/core/project_files.py

```diff
@@ -360,3 +386,68 @@ def _copy_yaml_attributes(  # noqa: WPS210
     def _write_file(self, file_path: PathLike, contents: Mapping):
         with atomic_write(file_path, overwrite=True) as fl:
             yaml.dump(contents, fl)
+
+    @classmethod
+    def extract_annotations(cls, data: CommentedMap | CommentedSeq) -> None:
+        """Extract annotations, and store them for later re-insertion.
+
+        Args:
+            data: The data from which to extract all "annotations" keys. It
+                will be modified in-place.
+        """
+        extracted = list(cls._extract_annotations(data=data, path=()))
+        if not cls._annotations:
+            cls._annotations = extracted
```

Comment:
@WillDaSilva - Looking more at internals here now that #6876 has shipped.... 🎉 

I'm sure you looked into this already and I think it would be helpful for me to better understand the context - and hopefully this is not a dumb question. 😅 Can you help me understand what was blocking the treatment of annotations as a property of the constructed classes which is simply ignored by them, but still serialized and deserialized in the same way as other properties?

I see we are extracting and restoring annotations but I guess I am missing context on why they cannot travel with the rest of the instantiated objects.

Reply:
@aaronsteers That was the first thing I tried, but I ruled it out because the way we build `Canonical` instances would require we deal with `annotations` at every level in the class hierarchy. Having every `Canoncial` subclass implement special handling for an attribute that `meltano` shouldn't ever touch was too messy:
- https://github.com/meltano/meltano/pull/7094

That said, I can probably give `Canonical` a metaclass that handles this before `__init__` gets called. It's invisible magic, which could trip someone up (possibly myself) in the future, but at least it contains the logic that handles annotations to a single location, and saves us from having to think about them in most cases.

I'll try that approach again now, and see what problem I run into. I wouldn't want to settle for a sub-par approach when a significantly better approach is possible.

Reply:
Thanks for asking about this. A second attempt at my first approach (this time using a metaclass) worked, and it's much simpler! I've implemented it in 86ef0e0.